code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/load/distance/measurers/gumtree/GumTreeDistanceMeasurer.kt

```diff
@@ -43,8 +47,11 @@ class GumTreeDistanceMeasurerByPsi(
     project: Project = createTempProject(),
     private val psiManager: PsiManager = PsiManager.getInstance(project),
 ) : GumTreeDistanceMeasurerBase() {
-    override fun String.parseTree(): TreeContext =
-        this.asPsiFile(Language.PYTHON, psiManager) {
-            getTreeContext(it)
+    private val codeToTreeContext: HashMap<String, TreeContext> = HashMap()
+    override fun String.parseTree(): TreeContext {
+        val trimmedCode = this.trimCode()
+        return codeToTreeContext[trimmedCode] ?: run {
+            this.asPsiFile(Language.PYTHON, psiManager) { getTreeContext(it) }
```

Comment:
Should we put new tree context into the hashmap here?

Reply:
oh yes, sorry